sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala

```diff
@@ -2412,4 +2412,29 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
       }
     }
   }
+
+  test("SPARK-29295: insert overwrite external partition should not have old data") {
+    withTable("test") {
+      withTempDir { f =>
+        sql("CREATE EXTERNAL TABLE test(id int) PARTITIONED BY (name string) STORED AS " +
+          s"PARQUET LOCATION '${f.getAbsolutePath}'")
+
+        withSQLConf(HiveUtils.CONVERT_METASTORE_PARQUET.key -> "false") {
+          sql("INSERT OVERWRITE TABLE test PARTITION(name='n1') SELECT 1")
```

Comment:
shall we fix the problem for dynamic partition as well? e.g. `INSERT TABLE ... PARTITION(p1=1,p2) ...`

Reply:
Dynamic partition insert also suffers the same issue. If we have consensus on deleting previous partition directory, I can also fix dynamic partition.

Reply:
since Hive already fixed in it the master branch, I think we should fix it as well.

Reply:
Ok. Would you prefer to fix dynamic partition in this PR too?

Reply:
Fixed dynamic partition too now.